docs(api protection): init protect your api doc (#59)
logto-io/docs: docs/docs/recipes/protect-your-api/README.md

sidebar_position: 2
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# ⚔️ Protect your API

Along with a fluent user sign-in experience, by default, Logto comes with the [API resources](../../references/resources/README.md) authorization service. It will help you protect the private APIs resources from anonymous identities. Let's walk through the following steps and implement the API resource authorization guard of your own using Logto.

:::info

> For more details, please check out our [API Resources doc](../../references/resources/README.md)

Author: Removed

> For more details, please check out our [API Resources doc](../../references/resources/README.md)

Author: removed

:::

> ## 1. Register the API resource through Logto **Admin Console**

Reviewer: maybe we don't need the prefix `1.` since they are listed in order on the right-side TOC

Author: updated

Logto will identify the registered [API resources](../../references/resources/README.md) from an authorization request and issue an audience-restricted `access_token` accordingly.

> First, go to the **API Resources** section in **Admin Console**. You will notice a build-in resource listed with the API identifier as `https://api.logto.io`. This resource indicates all the management APIs you may use to maintain Logto service. It will guarantee all our APIs are under protection and restricted to the Logto authorized users.

Reviewer: ```suggestion
Go to the **API Resources** section in **Admin Console**. You will notice a build-in resource listed with the API identifier as `https://api.logto.io`. This resource indicates all the management APIs you may use to maintain Logto service. It will guarantee all our APIs are under protection and restricted to the Logto authorized users with admin role.
```

Author: updated

<!-- TODO: Replace the API resource AC screenshot -->

> <img src="/img/docs/api_resource_landing.png" width="100%" />

Reviewer: the screenshot looks too big - can we focus on the header part of the page content?

> <img src="/img/docs/api_resource_landing.png" width="100%" />

Author: agreed just put a place holder here. Will review the screenshots with design and come with a more official version.

<br />
<br />

Next, click on the **Create API Resource** button to register your own API resource by providing:

> - A human-readable **API Name** that may better helps you to identify this entity later

Reviewer 2: ```suggestion
- A human-readable **API Name** that may better helps you to identify this entity later.
```

Author: updated

> - A human-readable **API Name** that may better helps you to identify this entity later

Author: Just remove the suffix I think is good though.

> - A unique **API Identifier** (a.k.a [Resource Indicator](../../references/resources/README.md#resource-indicator)). URI format variable that can represent the resource's identity you'd like to be guarded.

Reviewer 2: `be guarded` → `guard`?
```suggestion
- A unique **API Identifier** (a.k.a. [Resource Indicator](../../references/resources/README.md#resource-indicator)). URI format variable that can represent the resource's identity you'd like to guard.
```

Author: updated

<!-- TODO: Replace the API resource AC screenshot -->

<img src="/img/docs/api_resource_create.png" width="100%" />
<br />
<br />

:::caution

> The API Identifier is unique and used as the SSO of resource indicator for Logto. **NOT** editable once created. Be careful when you create it.

Author: updated

1. Registered in Logto server.
2. Initiated as configs into the Logto SDK at your application.

:::

<Tabs>
<TabItem value="kotlin" label="Kotlin">

```kotlin
logtoClient.getAccessToken("<your-target-api-resource>", { logtoException: LogtoException?, result: AccessToken? ->
  // custom logic
})
```

</TabItem>
<TabItem value="java" label="Java">

```java
logtoClient.getAccessToken("<your-target-api-resource>", (logtoException, accessToken) -> {
    // custom logic
});
```

</TabItem>
<TabItem value="swift" label="Swift">

```swift
  let token = try await client.getAccessToken(for: "<your-target-api-resource>")
  // custom logic
```

</TabItem>
<TabItem value="react" label="React">

```js
const { getAccessToken } = useLogto();
const token = await getAccessToken('<your-target-api-resource>');
// custom logic
```

</TabItem>

<TabItem value="vue" label="Vue">

```js
const { getAccessToken } = useLogto();
const token = await getAccessToken('<your-target-api-resource>');
// custom logic
```

</TabItem>

<TabItem value="js" label="VanillaJs">

```js
const token = await logtoClient.getAccessToken('<your-target-api-resource>');
// custom logic
```

</TabItem>

</Tabs>

> If the client is well-authorized, a JWT format `access_token` will be granted and issued by Logto, specifically for the requested entity. Encrypted, audience_restricted, and with lift-time. Carrying all the necessary info that can represent the authority of this request.

Reviewer 2: Should be `audience-restricted` and `Carry`?
```suggestion
If the client is well-authorized, a JWT format `access_token` will be granted and issued by Logto, specifically for the requested entity. Encrypted, audience-restricted, and with lift-time. Carry all the necessary info that can represent the authority of this request.
```

Reviewer 3: No, I think they are still in one sentence. `encrypted, audience-restricted and with life-time, carrying all the necessary info` are still modifying the preceding noun `access token`

Reviewer 3: Btw, there is a typo `lift-time` -> `life-time`, and `audience_based` should be `audience-based` with hyphen.

Reviewer 2: Maybe this sentence is too long to read. Grammarly may split it.

Author: Yep, use the whole sentence as an adj.  Intend to keep the sentence short to read.

Append the token to your request's Authorization header as the Bearer code:

i.e.

```bash
GET https://logto.dev/api/users

Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiYXVkIjoiaHR0cHM6Ly9sb2d0by5kZXYvYXBpL3VzZXJzIiwiaXNzIjoiaHR0cHM6Ly9sb2d0by5kZXYvb2lkYyIsIm5hbWUiOiJKb2huIERvZSIsImlhdCI6MTUxNjIzOTAyMiwiZXhwIjoxNTE2MzI1NDIyfQ.PjIJl00YNC84EPNYLEdpEEAdAxqhekCYhFEckvRokek

```

Now you have your application well equipped and ready to launch 🚀 .

## 3. Validate the Authorization header on your API side

//TODO:

> We follow the standard Code Based OAuth 2.0 Authorization Protocol. If you are interested in the strategy behind Logto, please refer to OAuth 2.0's [official document](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3.1) for more details.

Reviewer 2: nit: just personal preference 

I commonly see `See link.`, `See link for more.`, `See link for details.` or `See link for more details.` in the tech book.
```suggestion
We follow the standard Code Based OAuth 2.0 Authorization Protocol. If you are interested in the strategy behind Logto, see OAuth 2.0's [official document](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3.1) for more details.
```

Reviewer 3: `Please refer to` looks fine to me, tho

Author: depends on the sentence context.  would like to keep `refer to` here.
